File: 3573.best-time-to-buy-and-sell-stock-v.cpp
// @leet imports start
#include <bits/stdc++.h>
using namespace std;
// @leet imports end

// @leet start
class Solution {
    using ll = long long;

  private:
    vector<int> prices;
    vector<vector<vector<ll>>> memo;

    static constexpr int NOT_HOLDING = 0;
    static constexpr int HOLDING = 1;
    static constexpr int SHORT_SELL = 2;
    static constexpr ll NEG_INF = LLONG_MIN / 4;

    ll dp(int i, int k_left, int state, bool can_ss) {
        if (i == prices.size() || k_left == 0) {
            if (state == NOT_HOLDING)
                return 0;
            else
                return NEG_INF;
        }
        ll& ans = memo[i][k_left][state];
        if (ans != -1)
            return ans;

        if (state == NOT_HOLDING) {
            ll buy = -prices[i] + dp(i + 1, k_left, HOLDING, can_ss);
            if (can_ss) {
                ll shortsell =
                    prices[i] + dp(i + 1, k_left, SHORT_SELL, can_ss);
                ans = max(buy, shortsell);
            } else {
                ans = buy;
            }
        } else if (state == HOLDING) {
            ll sell = prices[i] + dp(i + 1, k_left - 1, NOT_HOLDING, can_ss);
            ans = sell;
        } else {
            // SHORT_SELL
            ll buy_back =
                -prices[i] + dp(i + 1, k_left - 1, NOT_HOLDING, can_ss);
            ans = buy_back;
        }
        ll skip = dp(i + 1, k_left, state, can_ss);
        ans = max(ans, skip);
        return ans;
    }

  public:
    long long maximumProfit(vector<int>& prices, int k) {
        this->prices = prices;
        memo.assign(prices.size(), vector(k + 1, vector<ll>(3, -1)));
        return dp(0, k, NOT_HOLDING, true);
    }
};
;
// @leet end

int main() {}

File: 3573.best-time-to-buy-and-sell-stock-v.py
# @leet imports start
from typing import *

# @leet imports end


# @leet start
from functools import cache


class Solution:
    def maximumProfit(self, prices: List[int], k: int) -> int:
        n = len(prices)

        NOT_HOLDING = 0
        HOLDING = 1
        SHORT_SELL = 2

        @cache
        def dp(i: int, k_left: int, state: int) -> int:
            if i == n or k_left == 0:
                if state == NOT_HOLDING:
                    return 0
                else:
                    return -(10**10)
            if state == NOT_HOLDING:
                buy = -prices[i] + dp(i + 1, k_left, HOLDING)
                skip = dp(i + 1, k_left, NOT_HOLDING)
                short = prices[i] + dp(i + 1, k_left, SHORT_SELL)
                return max(buy, skip, short)
            elif state == HOLDING:
                sell = prices[i] + dp(i + 1, k_left - 1, NOT_HOLDING)
                skip = dp(i + 1, k_left, HOLDING)
                return max(sell, skip)
            else:  # SHORT_SELL
                buy_back = -prices[i] + dp(i + 1, k_left - 1, NOT_HOLDING)
                skip = dp(i + 1, k_left, SHORT_SELL)
                return max(buy_back, skip)

        return dp(0, k, NOT_HOLDING)


# @leet end

p = [1, 7, 9, 8, 2]
k = 2
x = Solution().maximumProfit(p, k)
print(x)

[12, 16, 19, 19, 8, 1, 19, 13, 9]
3
